File: assiist_back_end/api/schemas/feedback.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field, validator

class FeedbackCreateRequest(BaseModel):
    """Request schema for creating feedback."""
    feedback_text: str = Field(..., min_length=1, max_length=5000, description="Feedback content")
    feedback_type: str = Field(default="general", description="Type of feedback")
    app_version: Optional[str] = Field(None, description="App version for debugging context")
    platform: Optional[str] = Field(None, description="Platform: ios, android, web")
    screen_context: Optional[str] = Field(None, description="Screen where feedback was submitted")

    @validator('feedback_type')
    def validate_feedback_type(cls, v):
        allowed_types = {'general', 'bug_report', 'feature_request'}
        if v not in allowed_types:
            raise ValueError(f'feedback_type must be one of: {", ".join(allowed_types)}')
        return v

    @validator('platform')
    def validate_platform(cls, v):
        if v is not None:
            allowed_platforms = {'ios', 'android', 'web'}
            if v not in allowed_platforms:
                raise ValueError(f'platform must be one of: {", ".join(allowed_platforms)}')
        return v

class FeedbackResponse(BaseModel):
    """Response schema for feedback operations."""
    id: str
    account_id: str
    user_id: str
    feedback_text: str
    feedback_type: str
    user_email: str
    user_name: str
    app_version: Optional[str] = None
    platform: Optional[str] = None
    screen_context: Optional[str] = None
    status: str
    created_on: datetime
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }

class FeedbackListResponse(BaseModel):
    """Response schema for listing feedback."""
    feedback: list[FeedbackResponse]
    total_count: int

class FeedbackUpdateRequest(BaseModel):
    """Request schema for updating feedback (admin use)."""
    status: Optional[str] = None
    admin_notes: Optional[str] = None

    @validator('status')
    def validate_status(cls, v):
        if v is not None:
            allowed_statuses = {'submitted', 'reviewed', 'resolved'}
            if v not in allowed_statuses:
                raise ValueError(f'status must be one of: {", ".join(allowed_statuses)}')
        return v 

File: assiist_front_end/lib/core/models/feedback.dart
class Feedback {
  final String id;
  final String accountId;
  final String userId;
  final String feedbackText;
  final String feedbackType;
  final String userEmail;
  final String userName;
  final String? appVersion;
  final String? platform;
  final String? screenContext;
  final DateTime createdOn;
  final String status;

  const Feedback({
    required this.id,
    required this.accountId,
    required this.userId,
    required this.feedbackText,
    required this.feedbackType,
    required this.userEmail,
    required this.userName,
    this.appVersion,
    this.platform,
    this.screenContext,
    required this.createdOn,
    required this.status,
  });

  factory Feedback.fromJson(Map<String, dynamic> json) {
    return Feedback(
      id: json['id'] as String,
      accountId: json['account_id'] as String,
      userId: json['user_id'] as String,
      feedbackText: json['feedback_text'] as String,
      feedbackType: json['feedback_type'] as String,
      userEmail: json['user_email'] as String,
      userName: json['user_name'] as String,
      appVersion: json['app_version'] as String?,
      platform: json['platform'] as String?,
      screenContext: json['screen_context'] as String?,
      createdOn: DateTime.parse(json['created_on'] as String),
      status: json['status'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'account_id': accountId,
      'user_id': userId,
      'feedback_text': feedbackText,
      'feedback_type': feedbackType,
      'user_email': userEmail,
      'user_name': userName,
      'app_version': appVersion,
      'platform': platform,
      'screen_context': screenContext,
      'created_on': createdOn.toIso8601String(),
      'status': status,
    };
  }
}

class FeedbackSubmissionRequest {
  final String feedbackText;
  final String feedbackType;
  final String? appVersion;
  final String? platform;
  final String? screenContext;

  const FeedbackSubmissionRequest({
    required this.feedbackText,
    this.feedbackType = 'general',
    this.appVersion,
    this.platform,
    this.screenContext,
  });

  Map<String, dynamic> toJson() {
    return {
      'feedback_text': feedbackText,
      'feedback_type': feedbackType,
      if (appVersion != null) 'app_version': appVersion,
      if (platform != null) 'platform': platform,
      if (screenContext != null) 'screen_context': screenContext,
    };
  }
}

class FeedbackListResponse {
  final List<Feedback> feedback;
  final int totalCount;

  const FeedbackListResponse({
    required this.feedback,
    required this.totalCount,
  });

  factory FeedbackListResponse.fromJson(Map<String, dynamic> json) {
    final feedbackList =
        (json['feedback'] as List)
            .map((item) => Feedback.fromJson(item as Map<String, dynamic>))
            .toList();

    return FeedbackListResponse(
      feedback: feedbackList,
      totalCount: json['total_count'] as int,
    );
  }
}
